### DISCUSSION TITLE
Not sure how to get tracing ingesting to work

### DISCUSSION THREAD

**Discussion by uhfath** [datalust/seq-tickets#2194] [Help] [answered]
Here is a sample of a log sent through `Vector` to `Seq` using HTTP ingesting:
```json
{
  "@t": "2024-05-14T08:09:37.5882985Z",
  "@mt": "Sending file. Request path: '{VirtualPath}'. Physical path: '{PhysicalPath}'",
  "@tr": "c249130babf90db839df1265014b1ce2",
  "@sp": "91c53fab6829c37e",
  "VirtualPath": "/_content/Radzen.Blazor/fonts/SourceSansPro-Semibold.woff",
  "PhysicalPath": "D:\\Projects\\MED.Framework\\ThirdParty\\radzen-blazor\\Radzen.Blazor\\wwwroot\\fonts\\SourceSansPro-Semibold.woff",
  "EventId": {
    "Id": 2,
    "Name": "FileServed"
  },
  "SourceContext": "Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware",
  "RequestId": "0HN3JTKN4UK22:00000007",
  "RequestPath": "/_content/Radzen.Blazor/fonts/SourceSansPro-Semibold.woff",
  "ConnectionId": "0HN3JTKN4UK22",
  "@sc": "Microsoft.AspNetCore.Hosting.HttpRequestIn",
  "@Elapsed": "00:00:00",
  "@st": "2024-05-14T08:09:37.5858124Z",
  "@ps": "0000000000000000",
  "Environment": "Development",
  "Application": "MED.Incident.UI.Host",
  "Assembly": "MED.Incident.UI.Host",
  "AssemblyVersion": "1.0.0.0"
}
```
And here is how it looks like in `Seq`:
![image](https://github.com/datalust/seq-tickets/assets/14792749/1a3e1702-ec01-4a75-8c7b-5e28bc0c3fe9)

There are a few things that I don't understand:
1. If I click `Trace` -> `Show` then only one single trace is shown in the window above. I thought there will be a series of traces. Also no matter how many times I click `Trace` -> `Show` on other entries, only the first one is shown - the window is not updated. Only disabling/enabling tracing with a button on the top helps to reset it.
2. What is the proper `CLEF` format to ingest `@Elapsed` as described [here](https://docs.datalust.co/docs/tracing#traces-in-seq). Should I just send it as `@Elapsed` or there is a special name for this key?
3. How to show the traces in a hierarchical manner like it's shown in `Seq` [docs](https://docs.datalust.co/docs/tracing#viewing-a-trace)?
4. Somehow `@sc` is shown as part of `Properties`, but was sent at the root level. Is something wrong with my log format? Same with `@Elapsed`.

If it helps, I've attached the log file (with just these entries):
[logs.log](https://github.com/datalust/seq-tickets/files/15305584/logs.log)

Can someone please explain to me how to use tracing properly? I'm sure there's something obvious I'm not getting, but I just can't figure it out.

**Comment by nblumhardt** [comment]
Hi!

1. The window will show all spans with the same trace id (`@tr`) - are you expecting there to be more spans in that trace?
2. `@Elapsed` isn't accepted on ingest; it's calculated as `@t - @st`
3. This falls out of (1), spans in the same trace need to have the same trace id, and the hierarchy is defined by their parent span (`@ps`) properties
4. `@sc` should be an object, like `"@sc": {"name": "..."}` - though it'll still appear in the properties list, just slightly differently :-)

If you're not running 2024.3, you'll need that in order for `@sc` to be recognized IIRC.

Hope this helps!

Nick

**Reply by uhfath** [reply]
Thank you for detailed information.
> The window will show all spans with the same trace id (@tr) - are you expecting there to be more spans in that trace?

Yes. All these entries use the same `@tr` value. The log file attached is "filtered" to these. It's also visible in a screenshot (the search field part).


> @Elapsed isn't accepted on ingest; it's calculated as @t - @st

Got it. Thanks! I've mistakenly thought that it should be also transmitted.


> This falls out of (1), spans in the same trace need to have the same trace id, and the hierarchy is defined by their parent span (@ps) properties

As to `@ps` it's also same for these entries - it's zero actually.


> @sc should be an object, like "@sc": {"name": "..."} - though it'll still appear in the properties list, just slightly differently :-)

Thanks. Will convert and take a look.

And yes, we are using `2024.3.11510` version right now.

**Reply by uhfath** [reply]
I've just recorded a small video clip of how it looks like on my machine.
Notice that trace window shows one specific entry and not all of them which are currently filtered under same `@tr`.

https://github.com/datalust/seq-tickets/assets/14792749/652d56c8-f67b-4fdc-80cc-246968b23756

Could this be because all entries have the same `@sp` ?

**Reply by liammclennan** [reply, marked as answer]
@uhfath `@sp` is the span id, so yes the behavior is undefined if many spans have the same span id.

**Reply by nblumhardt** [reply]
Just to add one more note, leaning on https://github.com/serilog-tracing/serilog-tracing for this should make everything much smoother. HTH!

**Reply by uhfath** [reply]
@liammclennan makes sense. Thanks!

@nblumhardt thanks for the link, but unfortunately if I understand correctly it works solely with Serilog's `ILogger` and not MEL which is required in our case. That is why we made our own enricher and some custom utility classes for activities.

**Reply by nblumhardt** [reply]
@uhfath no, SerilogTracing hooks into `System.Diagnostics.ActivitySource`, so if you create `Activity` instances in your code, you can collect them with SerilogTracing whether you are using its Serilog extensions or not. Your use case is one of the core scenarios it supports 👍  HTH!

**Reply by uhfath** [reply]
@nblumhardt interesting enough, I couldn't make it to work the first time I tried it but gave it another go after your comment and it seems to work now. Thanks!
But there is another issue now:
![image](https://github.com/datalust/seq-tickets/assets/14792749/003d74c5-2a8a-49e5-b302-4afd12bb998e)
Notice that "Missing" part. What could it be?
The setup is quite simple in my case. The only change is this:
```c#
using var _ = new ActivityListenerConfiguration()
    .Instrument.AspNetCoreRequests()
    .Instrument.HttpClientRequests()
    .TraceToSharedLogger();
```
Currently we are not using Serilog's activity extensions. Just this setup and standard MEL stuff.

**Reply by nblumhardt** [reply]
`Instrument.HttpClientRequests()` is the default so you shouldn't need it here (this will potentially create duplicate spans).

Are these spans from the Blazor app, or do they represent work going on in ASP.NET Core? Debugging missing trace spans is not a lot of fun, unfortunately - you'll likely need to dig around, put breakpoints at the places the spans above are created, and try to work out what code (or service) higher on the call stack is generating a span that's not captured.

Alternatively, if the caller isn't something you're in control of, using `Instrument.AspNetCoreRequests(opts => opts.IncomingTraceparent = IncomingTraceparent.Ignore)` will break these into to separate one-span traces, assuming that they're server-side.

HTH,
Nick

**Reply by uhfath** [reply]
> Instrument.HttpClientRequests() is the default so you shouldn't need it here (this will potentially create duplicate spans).

Thanks for the tip. Removed.

> Are these spans from the Blazor app, or do they represent work going on in ASP.NET Core?

These come from the ASP.NET Core API application.
There are, actually, two apps involved - one hosts Blazor app (operates as a simple web server for static files) and the other one is the API with one single endpoint. Blazor app code is currently not logged.

The screenshot showed an operation of querying one single endpoint from one app to another. It perfectly "connected" the traces but showed that "missing" thing. There is no user code that creates any activities. Just ASP.NET Core built-in methods.

If I use `IncomingTraceparent.Ignore` then traces are broken - i.e. no "connection" between client and server. I suppose that's because API part now ignores traces from HTTP request and creates it's own. The "missing" thing is also gone, but not sure if it's helpful in this case.

**Reply by nblumhardt** [reply]
Do you have a level override set for `Microsoft.AspNetCore`? If so, that could be filtering out the activity generated by ASP.NET.

**Reply by uhfath** [reply]
Nope. I have only default level set to "Verbose".
But I'm afraid I have messed up something when switched to `SerilogTacing` because there were some "traces" left of a previous enricher which we used. Will clean things up and get back to you.
